### PR TITLE
Dennisme/rebase pod events

### PR DESCRIPTION
**Pull request by dennisme**
**Issue #635, if available:**

**Description of changes:**
This PR rebases an [earlier PR](https://github.com/newrelic-forks/aws-node-termination-handler/pull/1) on our fork by @dthorsen that adds  pod eventing to draining. We have been running this for a few months and it does help a ton with visibility. 

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

**Comment by dennisme** [comment]
Tests failing locally 

```
2022/04/21 18:46:09 INF Error when trying to list Nodes w/ label, falling back to direct Get lookup of node
2022/04/21 18:46:09 INF Pods on node node_name=e2e-test-7621cf83-worker pod_names=["e2e-test-7621cf83-aemm-amazon-ec2-metadata-mock-5445f44f8fjk9zw","regular-pod-test-b5fb5fc6-jfttf","webhook-test-proxy-hdm49","e2e-test-7621cf83-anth-aws-node-termination-handler-j5zkb","kindnet-mcj95","kube-proxy-vhc7n"]
2022/04/21 18:46:09 INF Error when trying to list Nodes w/ label, falling back to direct Get lookup of node
2022/04/21 18:46:09 INF Draining the node
2022/04/21 18:46:09 INF Error when trying to list Nodes w/ label, falling back to direct Get lookup of node
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x18 pc=0x1b39097]

goroutine 113 [running]:
github.com/aws/aws-node-termination-handler/pkg/observability.(*K8sEventRecorder).AnnotatedEventf(0xc000848200, 0x24d7808, 0xc00044fc70, 0xc000873350, 0x21a2e27, 0x6, 0x21a8219, 0xb, 0x21c484b, 0x1d, ...)
	<autogenerated>:1 +0x37
github.com/aws/aws-node-termination-handler/pkg/node.Node.CordonAndDrain(0x0, 0xc00004003a, 0x18, 0xc00003e059, 0x39, 0xc000048016, 0x46, 0x101, 0xc0000401f8, 0x9, ...)
	/node-termination-handler/pkg/node/node.go:140 +0x515
main.cordonAndDrainNode(0x0, 0xc00004003a, 0x18, 0xc00003e059, 0x39, 0xc000048016, 0x46, 0x101, 0xc0000401f8, 0x9, ...)
	/node-termination-handler/cmd/node-termination-handler.go:404 +0x12b
main.drainOrCordonIfNecessary(0xc000256240, 0xc0006741e0, 0x0, 0xc00004003a, 0x18, 0xc00003e059, 0x39, 0xc000048016, 0x46, 0x101, ...)
	/node-termination-handler/cmd/node-termination-handler.go:354 +0x665
created by main.main
	/node-termination-handler/cmd/node-termination-handler.go:247 +0x1107
⏰ Took 988sec
```

**Comment by brycahta** [comment]
@dennisme thanks for the PR. Could you include the issue # or requested feature in the description? If no issue exists, please open one with some context as to why these changes are needed/beneficial.

**Comment by dennisme** [comment]
> @dennisme thanks for the PR. Could you include the issue # or requested feature in the description? If no issue exists, please open one with some context as to why these changes are needed/beneficial.

Yes of course! Also next commit will fix the test 😬

**Comment by dennisme** [comment]
ok we should be good, test fixed :)

**Review by brycahta — CHANGES_REQUESTED** [review]

File: pkg/node/node.go [hunk 1]
```diff
@@ -64,6 +65,13 @@ const (
 	maxTaintValueLength = 63
 )
 
+const (
+	// PodEvictReason is the event reason emitted for Pod evictions during node drain
+	PodEvictReason = "PodEviction"
+	// PodEvictMsg is the event message emitted for Pod evictions during node drain
+	PodEvictMsg = "Pod evicted due to node drain"
```
Comment (snay2) [inline pkg/node/node.go:72]: Is this sufficient detail for the event? Do you need it to specify that node-termination-handler initiated the drain in response to a particular AWS event (Spot ITN, ASG event, etc.)?

File: pkg/node/node.go [hunk 2]
```diff
@@ -114,6 +122,25 @@ func (n Node) CordonAndDrain(nodeName string, reason string) error {
...
+					Kind:      "Pod",
+					Name:      pod.Name,
+					Namespace: pod.Namespace,
+				}
+				annotations := make(map[string]string)
+				annotations["node"] = nodeName
+				for k, v := range pod.GetLabels() {
+					annotations[k] = v
+				}
+				recorder.AnnotatedEventf(podRef, annotations, "Normal", PodEvictReason, PodEvictMsg)
```
Comment (snay2) [inline pkg/node/node.go:140]: Should `"Normal"` also be in a constant, like `PodEvictReason` and `PodEvictMsg`? What are the other possible values for that argument?

File: pkg/observability/k8s-events.go [hunk 3]
```diff
@@ -113,7 +113,7 @@ func InitK8sEventRecorder(enabled bool, nodeName string, sqsMode bool, nodeMetad
 	}
 
 	broadcaster := record.NewBroadcaster()
-	broadcaster.StartRecordingToSink(&typedcorev1.EventSinkImpl{Interface: clientSet.CoreV1().Events("default")})
+	broadcaster.StartRecordingToSink(&typedcorev1.EventSinkImpl{Interface: clientSet.CoreV1().Events("")})
```
Comment (snay2) [inline pkg/observability/k8s-events.go:116]: Can you provide some context on this change? What is the difference in behavior between `"default"` and `""`?

File: pkg/node/node.go [hunk 4]
```diff
@@ -800,3 +827,7 @@ func filterPodForDeletion(podName string) func(pod corev1.Pod) drain.PodDeleteSt
 		return drain.MakePodDeleteStatusOkay()
 	}
 }
+
+type recorderInterface interface {
+	AnnotatedEventf(object runtime.Object, annotations map[string]string, eventtype, reason, messageFmt string, args ...interface{})
```
Comment (AustinSiu) [inline pkg/node/node.go:832]: nit: `eventtype` -> `eventType`

File: pkg/node/node_test.go [hunk 5]
```diff
@@ -61,7 +63,8 @@ func TestDryRun(t *testing.T) {
 	tNode, err := node.New(config.Config{DryRun: true})
 	h.Ok(t, err)
 
-	err = tNode.CordonAndDrain(nodeName, "cordonReason")
+	fakeRecorder := record.NewFakeRecorder(100)
```
[the remark this answers is not in the document]
Comment (dennisme) [reply pkg/node/node_test.go:66]: Probably not, but if it set to small the tests will hang indefinitely.

File: pkg/node/node_test.go [hunk 6]
```diff
@@ -98,6 +101,7 @@ func TestNewFailure(t *testing.T) {
 }
 
 func TestDrainSuccess(t *testing.T) {
+	controllerBool := true
```
[the remark this answers is not in the document]
Comment (dennisme) [reply pkg/node/node_test.go:104]: Good call

File: pkg/node/node.go (hunk 4 above)
Comment (dennisme) [reply pkg/node/node.go:832]: Yes good catch. I agree.

**Pull request closed without merging**